Use Terraform for deploying lambdas (#556) (#1635)
DataBiosphere/azul: UPGRADING.rst

Upgrading
---------

This file documents any upgrade procedure that must be performed. Because we
don't use a semantic version, a change that requires explicit steps to upgrade
a is referenced by its Github issue number. After checking out a branch that
contains a commit resolving an issue listed here, the steps listed underneath
the issue need to be performed. When switching away from that branch, to a
branch that does not have the listed changes, the steps would need to be
reverted. This is all fairly informal and loosely defined. Hopefully we won't
have too many entries in this file.

#558 Deploying lambdas with Terraform
=====================================

Switching **TO** this branch

> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Reviewer: ```suggestion
```

No one will switch to "this branch". It isn't even clear what "this branch" refers to. Most people will switch to a branch that includes this *change* and then they'll rebase the rest of their branches. So in essence, they'll do this once and hardly anyone will have to downgrade.

Author: Understood. I'll remove this heading.

To deploy Lambdas with Terraform you will need to remove the currently deployed

> Lambda resources from Chalice. Checkout a commit *before* these changes and run

Reviewer: ```suggestion
Lambda resources from Chalice. Checkout the most recent commit *before* these changes and run
```

::

    make -C lambdas delete

> Switch back to this branch. Now use Chalice to generate the new Terraform

Reviewer: another reference to "this branch"

config. Run ::

    make deploy

And finally ::

    make terraform

Switching **FROM** this branch

> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Reviewer: ```suggestion
In the unlikely case that you need to downgrade, perform the steps below.
```

have too many entries in this file.

#558 Deploying lambdas with Terraform
=====================================

To deploy Lambdas with Terraform you will need to remove the currently deployed
Lambda resources from Chalice. Checkout the most recent commit *before* these
changes and run ::
    cd terraform
    make init
    terraform destroy $(terraform state list | grep aws_api_gateway_base_path_mapping | sed 's/^/-target /')
    cd ..
    make -C lambdas delete

Switch back to your branch that includes these changes. Now use Chalice to
generate the new Terraform config. Run ::

    make deploy

And finally ::

    make terraform

In the unlikely case that you need to downgrade, perform the steps below.

Switch to the new branch you want to deploy. Run ::

    cd terraform
    rm -r indexer/ service/
    make init
    terraform destroy $(terraform state list | grep aws_api_gateway_base_path_mapping | sed 's/^/-target /')
    cd ..
    make terraform

This will remove the Lambda resources provisioned by Terraform. Now run ::

    make deploy

to set up the Lambdas again, and finally ::

    make terraform

To complete the API Gateway domain mappings, etc.

Run ::

>     make deploy a final time out work around a bug with OpenAPI spec generation.

Reviewer 2: ```suggestion
Run ::

    make deploy

a final time out work around a bug with OpenAPI spec generation.
```

Upgrading
---------

This file documents any upgrade procedure that must be performed. Because we
don't use a semantic version, a change that requires explicit steps to upgrade
a is referenced by its Github issue number. After checking out a branch that
contains a commit resolving an issue listed here, the steps listed underneath
the issue need to be performed. When switching away from that branch, to a
branch that does not have the listed changes, the steps would need to be
reverted. This is all fairly informal and loosely defined. Hopefully we won't
have too many entries in this file.

#558 Deploying lambdas with Terraform
=====================================

To deploy Lambdas with Terraform you will need to remove the currently deployed
Lambda resources from Chalice. Checkout the most recent commit *before* these
changes and run ::

>     cd terraform

Author: It was an oversight, but thanks for the explanation.
